src/environment.cpp

```diff
@@ -116,6 +117,16 @@ void Environment::continueRaycast(RaycastState *state, PointedThing *result)
 {
 	const NodeDefManager *nodedef = getMap().getNodeDefManager();
 	if (state->m_initialization_needed) {
+		if (objectpos_over_limit(state->m_shootline.start) || objectpos_over_limit(state->m_shootline.end)) {
+			warningstream << "Environment::continueRaycast(): "
+				<< "raycast positions (" << state->m_shootline.start.X << "," << state->m_shootline.start.Y << "," << state->m_shootline.start.Z
+				<< " -> " << state->m_shootline.end.X << "," << state->m_shootline.end.Y << "," << state->m_shootline.end.Z
+				<< ") outside maximum range, won't raycast" << std::endl;
```

Comment:
IMO the warning here is not needed.

More importantly, should this really abort the entire raycast?
A raycast could start outside the map but still hit something valid. It might be better to clip the coordinates to the first point of the ray that is inside the map ([like this I guess?](https://webcache.googleusercontent.com/search?q=cache:UA4_BJV5AzcJ:https://www.scratchapixel.com/lessons/3d-basic-rendering/minimal-ray-tracer-rendering-simple-shapes/ray-box-intersection))

Reply:
Also: `PP(state->m_shootline.end)` would be way shorter to output the position in the usual format.

Reply:
Untested limiting code example, based on what sfan5 recommended.
```C++
if (objectpos_over_limit(state->m_shootline.start)) {
    // abort, no collisions
	return;
}

v3f extent = state->m_shootline.end - state->m_shootline.start;
f32 fraction_of_extent = 0; // Fraction of the extent to move the end position backwards

f32 overlap = std::abs(state->m_shootline.end.X) - MAX_MAP_GENERATION_LIMIT * BS;
if (overlap > 0) {
    // Limit by X. Re-add sign to overlap, get linear fraction of the distance within the extent cuboid
	fraction_of_extent = MYMAX(fraction_of_extent, (overlap * MY_SIGN(state->m_shootline.end.X)) / extent.X);
}

overlap = std::abs(state->m_shootline.end.Y) - MAX_MAP_GENERATION_LIMIT * BS;
if (overlap > 0) {
	fraction_of_extent = MYMAX(fraction_of_extent, (overlap * MY_SIGN(state->m_shootline.end.Y)) / extent.Y);
}

overlap = std::abs(state->m_shootline.end.Z) - MAX_MAP_GENERATION_LIMIT * BS;
if (overlap > 0) {
	fraction_of_extent = MYMAX(fraction_of_extent, (overlap * MY_SIGN(state->m_shootline.end.Z)) / extent.Z);
}

state->m_shootline.end -= fraction_of_extent * extent; // move it backwards
```

Reply:
Well warning is useful to catch griefers, but not totally necessary.

Reply:
>> IMO the warning here is not needed.
> Well warning is useful to catch griefers, but not totally necessary.

From a game or mod dev standpoint the warning is very useful to catch coding errors, not only to catch griefers.

> A raycast could start outside the map but still hit something valid.

This is exactly the kind of bug that is more likely to happen if you stop stuff at mapgen limit btw. …

> It might be better to clip the coordinates to the first point of the ray that is inside the map

Correction: Clip the coordinates to the first point of the ray that may fit inside s16 limits.